libbeat/publisher/queue/memqueue/runloop.go

```diff
@@ -174,59 +200,180 @@ func (l *runLoop) getRequestShouldBlock(req *getRequest) bool {
+		deletedByteCount += l.buf.entry(entryIndex).eventSize
+	}
+	l.bufferStart = l.bufferStart.plus(deletedEntryCount)
+	l.eventCount -= deletedEntryCount
+	l.byteCount -= deletedByteCount
+	l.consumedEventCount -= deletedEntryCount
+	l.consumedByteCount -= deletedByteCount
+	l.observer.RemoveEvents(deletedEntryCount, deletedByteCount)
 	if l.closing && l.eventCount == 0 {
 		// Our last events were acknowledged during shutdown, signal final shutdown
 		l.broker.ctxCancel()
 	}
+
+	// We just freed up space in the queue, see if this unblocked any
+	// pending inserts.
+	l.maybeUnblockPushRequests()
+}
+
+func (l *runLoop) handlePushRequest(req pushRequest) {
+	// If other inserts are already pending, or we don't have enough room
+	// for the new entry, we need to either reject the request or block
+	// until we can handle it.
+	if !l.pendingPushRequests.Empty() || !l.canFitPushRequest(req) {
+		if req.blockIfFull {
+			// Add this request to the pending list to be handled when there's space.
+			l.pendingPushRequests.Add(req)
+		} else {
+			// Request doesn't want to block, return failure immediately.
+			l.broker.logger.Debugf("queue is full, dropping event")
+			req.resp <- false
+		}
+		return
+	}
+	// There is space, insert the new event and report the result.
+	l.doInsert(req)
+}
+
+// Returns true if the given push request can be added to the queue
+// without exceeding the entry count or byte limit.
+func (l *runLoop) canFitPushRequest(req pushRequest) bool {
+	if l.broker.useByteLimits() {
+		newByteCount := l.byteCount + req.eventSize
+		return newByteCount <= l.broker.settings.Bytes
+	}
+	newEventCount := l.eventCount + 1
+	return newEventCount <= l.broker.settings.Events
 }
 
-func (l *runLoop) handleInsert(req *pushRequest) {
-	l.insert(req, l.nextEntryID)
-	// Send back the new event id.
-	req.resp <- l.nextEntryID
+func (l *runLoop) maybeUnblockPushRequests() {
+	for !l.pendingPushRequests.Empty() {
+		req := l.pendingPushRequests.First()
+		if l.closing {
+			// If the queue is closing, reject all pending requests
+			req.resp <- false
+			continue
+		}
+		if !l.canFitPushRequest(req) {
+			// We're out of space, the rest of the blocked requests will have
+			// to wait.
+			break
+		}
+		l.doInsert(req)
+		l.pendingPushRequests.Remove()
+	}
+}
+
+// growEventBuffer is called when there is no limit on the queue event
+// count (i.e. the queue size is byte-based) but the queue's event buffer
+// (a []queueEntry) is full.
+// For this to be possible, queue indices must be stable when the buffer
+// size changes. Therefore, entry positions are based on a strictly
+// increasing id, so that different events have different positions,
+// even when they occupy the same location in the underlying buffer.
+// The buffer position is the entry's index modulo the buffer size: for
+// a queue with buffer size N, the entries stored in buf[0] will have
+// entry indices 0, N, 2*N, 3*N, ...
+func (l *runLoop) growEventBuffer() {
```

Comment:
This seems reasonable but seems like a good reason to introduce benchmark tests

Reply:
Benchmarking `growEventBuffer` specifically? We could do that, although (since buffer growth is one-way) we expect this to be called a ~constant number of times on any run. (If the settings and event sizes are such that the queue needs to store a million events simultaneously, that's still only 10 calls to `growEventBuffer` over the lifetime of the program, the main bottlenecks are still in actually processing the events.)

Reply:
Not necessarily growEventBuffer, I just wanted us to stop and evaluate if there was any place microbenchmarking would be beneficial and save us time going through iterations of the end to end performance framework. If the answer is they aren't beneficial that is fine.

I think in this case the most impactful choice is probably the starting size of the buffer since that defines how many initial doublings need to happen to get to whatever steady state is.